client/blb/client.go

```diff
@@ -1127,109 +1126,123 @@ func (cli *Client) readOneTractReplicated(
-			for i := read; i < len(thisB); i++ {
-				thisB[i] = 0
+			result := readFunc(host, n, thisBLoc)
+
+			ch <- tractResultRepl{
+				tractResult: result,
+				badReadHost: host,
+				thisB:       thisBLoc,
 			}
-			ch <- tractResult{len(thisB), read, err, badVersionHost}
-			return
 		}
 
-		// Backup request logic involves the following:
-		// 1) Send the first read to the first host (item 0 available from the orderCh) in a goroutine.
-		// 2) Start another goroutine to issue backup requests. Each subsequent request is spawned after the
-		// configured backup time.After delay time expires.
-		// 3) Block until the first read is returned from the goroutines servicing the reads. Note that if the
-		// the request returned fails, we accept that and continume on to the fallback phase for sequential reads.
-		// We could optimize this later and scan for the first successful backup request, but that may not be
-		// worth the latency savings.
+		var wg sync.WaitGroup
+		nb := min(maxNumBackups+1, len(order))
+		wg.Add(nb)
 		go func() {
-			nb := min(maxNumBackups+1, len(order))
-			for i := 0; i < nb; i++ {
+			for n := 0; n < nb; n++ {
 				select {
-				case <-cli.backupReadState.backupDelayFunc(time.Duration(i) * delay):
-					go readFunc(order[i])
+				case <-cli.backupReadState.delayFunc(time.Duration(n) * delay):
+					thisBLoc := make([]byte, len(thisB))
+					go readFuncBackup(&wg, n, thisBLoc)
 				case <-ctx.Done():
-					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, i)
+					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, n)
 					break
 				}
 			}
 		}()
+		go func() {
+			wg.Wait()
+			close(ch)
+		}()
 
-		// take the first result.
-		*result = <-ch
-		if result.err == core.NoError {
-			return
+		// Take the first result or mark bad hosts from being read during the
+		// fallback phase below.
+		for res := range ch {
+			if res.err == core.NoError {
+				*result = res.tractResult
+				copy(thisB, res.thisB)
+				return
+			}
+			badReadHosts[res.badReadHost] = true
 		}
 	}
 
-	// Sequentailly read from tract server replicas. This phase will run if backups are disabled from config state, or
+	// Sequentially read from tract server replicas. This phase will run if backups are disabled from config state, or
 	// the backup read phase fails and we use this as a fallback.
 	for n := range order {
 		host := tract.Hosts[n]
 		if host == "" {
 			log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[n])
 			continue
 		}
-
-		var read int
-		read, err = cli.tractservers.ReadInto(ctx, host, nil, reqID, tract.Tract, tract.Version, thisB, thisOffset)
-		if err == core.ErrVersionMismatch {
-			badVersionHost = host
-		}
-		if err != core.NoError && err != core.ErrEOF {
-			log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
-			// TODO(eric): redo bad TS reporting mechanism.
-			continue // try another host
+		if badReadHosts[host] {
```

Comment:
A map seems kind of overkill for this purpose. We know how many requests we did and where we did them to (e.g. order[0] and order[1]). So why not just start at order[2]?